### DISCUSSION TITLE
[Fast Data] Avoid generating Single View Triggers when no projection fields are updated

### DISCUSSION THREAD

**Discussion by lucascanna** [mia-platform/community#511] [Q&A] [answered]
### What product are you having troubles with?

Console

### What Console version are you using?

13.0.0

### Description

Hi!

I have a projection that includes only a subset of the fields from the original database. As a result, Kafka messages are generated on the ingestion topic even when there are no relevant changes in the original database.

I expect the Projection Storer to recognize that none of the projection fields have been updated and, therefore, not generate any pr-update messages. However, this doesn't seem to be happening, at least with my current configuration. Is there a way to configure it to work as expected?

Thanks!

**Comment by flower-of-the-bridges** [comment]
Hi @Lucascanna 

The projection storer does not perform a check on before-after content. It just performs a check on the type of ingestion operation and performs either insert/update/delete operations on MongoDB.

Then a `pr-update` is generated.

What you can do is define [a custom function](https://docs.mia-platform.eu/docs/fast_data/configuration/config_maps/kafka_projection_updates) on the single view trigger reading those pr-updates to perform checks on before-after content.

Since a custom strategy accepts the whole message value as second parameter, you can do something like this:

```js
module.exports = async function* myCustomStrategy ({ logger, dbMongo }, { operation: {type: operationType}, before, after }) {
  
  if(operationType === 'INSERT') {
   // if is a insert operation, just yield
    yield { IDENTIFIER_FIELD: after.FIELD }
  } else if(operationType === 'UPDATE' && before.FIELD_TO_CHECK !== after.FIELD_TO_CHECK) {
    // I generate an identifier only if FIELD_TO_CHECK changes between before and after
    yield { IDENTIFIER_FIELD: after.FIELD }
  }
}
```

In this way, you would still generate `pr-update` messages, but no `sv-trigger` messages would be generated from the event if the values of interest do not change: does this solution help your use case or there's a need to not generate any `pr-update` event at all?

**Reply by lucascanna** [reply]
Hi @flower-of-the-bridges!

Thanks for your prompt reply.

Although my goal is not specifically to avoid the `pr-update` event, I don't really like the solution you're proposing because it requires me to replicate the data model of the projection within this function. I would have to create a custom strategy for each projection and add an array of `FIELDS_TO_CHECK` in each strategy. This is not desirable.

I would prefer to write the data model only once. That's why I believe this task should be handled by the Projection Storer, which is already aware of the data model. Do you have any plans to implement this feature? I see a very similar request [here](https://github.com/mia-platform/community/discussions/484).

Regards.

**Reply by flower-of-the-bridges** [reply]
Hi @Lucascanna I agree with you, as the number of projection increases (or just the fields!), this solution is not feasible for large-scale projects.

Currently, as you said, only the PS knows fast data models, since the trigger microservice does not load fast data models to infer this kind of update assumption.

I agree with you that this feature would help a lot of similar use cases, we will keep track of it with @albertotessarotto to understand how to plan it in our roadmap

Bests
Giovanni

**Reply by albertotessarotto** [reply, marked as answer]
Thanks for the issue. I agree about not having to rewrite data models multiple times, an activity that can also lead to errors. Regarding the reported use case, I don't know if it is possible to make changes at the CDC level to mitigate this scenario; from the product perspective, in conjunction with the general improvement of the UX of fast data configuration, we will plan the improvement for this type of use case by the end of the year and will keep you updated.

**Reply by lucascanna** [reply]
Hi @albertotessarotto! 

Yes, modifying the CDC configuration is a viable solution and it is the one we plan to implement. However, I want to point out that we are still duplicating the projection data model. In our case, using Debezium, modifying the CDC configuration involves adding the following property to the Debezium configuration:
```
columns.include.list: <all the columns that are involved in the projections>
```
As you can see, this results in duplicating the projection data model within the Debezium configuration.

I am happy to see this issue will be solved in the next Fast Data releases.

Thanks
